sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala

```diff
@@ -50,8 +50,25 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
   override def conf: SQLConf = SQLConf.get
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
-    case p: Union if p.children.forall(isEmptyLocalRelation) =>
-      empty(p)
+    case p @ Union(children) if children.exists(isEmptyLocalRelation) =>
+      val newChildren = children.filterNot(isEmptyLocalRelation)
+      if (newChildren.isEmpty) {
+        empty(p)
+      } else {
+        val newFirstChild = if (newChildren.head eq children.head) {
```

Comment:
I think you can use `sameOutput` like this;
```
      if (newChildren.isEmpty) {
        empty(p)
      } else {
        val newPlan = if (newChildren.size > 1) Union(newChildren) else newChildren.head
        // Describes why we need this handling
        if (newPlan.sameOutput(p)) {
          newPlan
        } else {
          val outputAliases = newPlan.output.zip(p.output).map { case (newAttr, outputName) =>
            Alias(newAttr, outputName.name)(newAttr.exprId)
          }
          Project(outputAliases, newPlan)
        }
      }
```

Reply:
Seems to be a good idea, but `sameOutput` uses `semanticEquals` that fails if an expression is non-deterministic. How about just comparing the expression ids like this?
```
     if (newChildren.isEmpty) {
        empty(p)
      } else {
        val newPlan = if (newChildren.size > 1) Union(newChildren) else newChildren.head
        val outputs = newPlan.output.zip(p.output)
        // the original Union may produce different output attributes than the new one so we alias them if needed
        if (outputs.forall { case (newAttr, oldAttr) => newAttr.exprId == oldAttr.exprId }) {
          newPlan
        } else {
          val outputAliases = outputs.map { case (newAttr, oldAttr) =>
            Alias(newAttr, oldAttr.name)(oldAttr.exprId)
          }
          Project(outputAliases, newPlan)
        }
      }

```

Reply:
I've changed it to my version in the latest commit. @maropu please let me know if yours with `sameOutput` would be better.

Reply:
Yea, your version looks okay.